aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -207,6 +207,144 @@ struct vec_host_softmax_lastdim {
+              output_m256_o1 = output_m256_o1.exp();
+              output_m256_o2 = output_m256_o2.exp();
+
+              output_m256_o1.store(temp_vec_output_data + d*vectorized_step*2);
+              output_m256_o2.store(temp_vec_output_data + d*vectorized_step*2 + vectorized_step);
+
+              sum_m256_o1 = sum_m256_o1 + output_m256_o1;
+              sum_m256_o2 = sum_m256_o2 + output_m256_o2;
+            }
+            // Step3: Unify
+            for (int64_t d = 0; d < dim_size; d += 1) {
+              Vec output_m256_o1 = Vec::loadu(temp_vec_output_data + d*vectorized_step*2);
+              Vec output_m256_o2 = Vec::loadu(temp_vec_output_data + d*vectorized_step*2 + vectorized_step);
+              output_m256_o1 = output_m256_o1/sum_m256_o1;
+              output_m256_o2 = output_m256_o2/sum_m256_o2;
+              Vec_bf16 output_m256_bf16 = convert_float_bfloat16(output_m256_o1, output_m256_o2);
+              output_m256_bf16.store(output_data + d * dim_stride);
+            }
+            idx += vectorized_step;
+          } else {
+            // Tail case(Scalar): it is exactly same logic as host_softmax
+            // inside aten/src/ATen/native/SoftMax.cpp. There are 2 kind of
+            // cases which will fall through this part:
+            // Case 1: For the idx at the end of total chunk for each thread, there are not enough numbers for parallization.
+            // Case 2: For the idx at the end of each inner_size inside thread, there are not enough numbers for parallization.
+            int64_t tail_number = ((idx+vectorized_step) > end) ? /*Case1*/ (end - idx) : /*Case2*/ (inner_size - inner_idx);
+            for (int64_t i=0; i < tail_number; i++) {
+              outer_idx = (idx + i) / inner_size;
+              inner_idx = (idx + i) % inner_size;
+              BFloat16* input_data =
+                  input_data_base + outer_idx * outer_stride + inner_idx;
+              BFloat16* output_data =
+                  output_data_base + outer_idx * outer_stride + inner_idx;
+              // Step1: Get max score
+              temp_scalar_input[0] = float(input_data[0]);
+              float max_input = temp_scalar_input[0];
+              for (int64_t d = 1; d < dim_size; d += 1) {
+                temp_scalar_input[d] = float(input_data[d * dim_stride]);
+                max_input = std::max(max_input, temp_scalar_input[d]);
+              }
+              // Step2: Calculate the Sum
+              float sum_data = 0.0;
+              for (int64_t d = 0; d < dim_size; d += 1) {
+                temp_scalar_output[d] = std::exp(temp_scalar_input[d] - max_input);
+                sum_data += temp_scalar_output[d];
+              }
+              // Step3: Unify
+              for (int64_t d = 0; d < dim_size; d += 1) {
+                output_data[d * dim_stride] =
+                    c10::BFloat16(temp_scalar_output[d]/sum_data);
+              }
```

Comment:
re: the allocations, I would be sort of inclined to try removing `temp_scalar_input` entirely, like this:

```
float max_input = -std::numeric_limits<float>::infinity();
for (int64_t d = 0; d < dim_size; d++) {
  max_input = std::max(max_input, input_data[d * dim_stride]);
}
// Step2: Calculate the Sum
float sum_data = 0.0;
for (int64_t d = 0; d < dim_size; d++) {
  temp_scalar_output[d] = std::exp(input_data[d * dim_stride] - max_input);
   sum_data += temp_scalar_output[d];
}
```

(it's also unusual in C++ to write `d += 1` instead of `d++`).

The reasoning for this being that using less memory is generally faster by way of being less taxing on the CPU cache and that I would hope the processor's stride prefetcher would help hide the latency of memory access to `input_data`. However, as an Intel employee, you've probably got access to much better information on that than I do. :)

a plain old heap allocation with std::unique_ptr seems better than a Tensor to me -- they're both going to do heap allocations, but Tensors have dispatcher overhead too and no benefit that I'm aware of if you're just going to use the data_ptr.

Reply:
Yean, agree. I would remove `temp_scalar_input ` to make the code tidy. 
The reason I use the `temp_scalar_input` is trying to coalesce 2 implicit BF16<->FP32 cast into 1. Like this in your example:

- `max_input = std::max(max_input, input_data[d * dim_stride]);`
- `temp_scalar_output[d] = std::exp(input_data[d * dim_stride] - max_input);` 

The `input_data[d * dim_stride]` is BF16, and would be casted to FP32 twice for calculation.
Generally, I think it can benefit the performance and accuracy. But it also makes the code complicated.

Reply:
So, I should remove `temp_scalar_input` and `temp_scalar_output`. And I will keep `temp_vec_input` and `temp_vec_output` in plain old heap allocation with std::unique_ptr.

Reply:
changed. @cpuhrsch @swolchok please help to review again.